sqlest/src/main/scala/sqlest/executor/Database.scala

```diff
@@ -207,6 +212,54 @@ case class Transaction(database: Database) extends Session(database) {
       } catch { case e: SQLException => }
     }
 
+  private[sqlest] def runAsync[A](f: Transaction => Future[A])(implicit ec: ExecutionContext): Future[A] = {
+    val runResult = try {
+      connection.setAutoCommit(false)
+      val connectionResult = try {
+        var success = false
+
+        val functionResult = try {
+          val functionCall = f(this)
+          functionCall andThen {
+            case Success(_) =>
+              if (shouldRollback) connection.rollback
+              else connection.commit
+              success = true
+              functionCall
+            case Failure(_) =>
+              connection.rollback
+              functionCall
+          }
+        } catch {
+          case NonFatal(e) => Future.failed(e)
+        }
+
+        functionResult andThen {
+          case _ =>
+            if (!success) connection.rollback
+            functionResult
+        }
+      } catch {
+        case NonFatal(e) => Future.failed(e)
+      }
+
+      connectionResult andThen {
+        case _ =>
+          connection.setAutoCommit(true)
+          connectionResult
+      }
+    } catch {
+      case NonFatal(e) => Future.failed(e)
+    }
+    runResult andThen {
+      case _ =>
+        try {
+          connection.close
+        } catch { case e: SQLException => }
+        runResult
+    }
+  }
+
```

Comment:
I have mirrored the synchronous code pretty closely - it's ugly but presumably the Java API is very exception-heavy